kadena/solidity/contracts/src/Wrapper.sol

```diff
@@ -0,0 +1,164 @@
+    bytes proof;
+    bytes publicValues;
+    bytes32 vkey;
+}
+
+contract Wrapper is SphinxPlonkVerifier, Ownable(msg.sender) {
+    error ErrorUnexpectedLongestChainFixture();
+    error ErrorUnexpectedSpvFixture();
+    error ErrorUnexpectedSpvSubjectHash();
+    error ErrorConfirmationWorkIsSmallerThanThreshold();
+    error ErrorUnexpectedLayerHash();
+
+    // state
+    bytes32[] checkpoints;
+    uint256 confirmation_work_threshold;
+
+    constructor(uint256 confirmation_work_threshold_, bytes32[] memory checkpoints_) {
+        // Kadena's Chainweb protocol setting
+        require(checkpoints_.length == 5);
+
+        checkpoints = checkpoints_;
+        confirmation_work_threshold = confirmation_work_threshold_;
+    }
+
+    // FIXME: This function is only for testing purposes!!!
+    function set_head_checkpoint(bytes32 zero_checkpoint) public onlyOwner {
+        checkpoints[0] = zero_checkpoint;
+    }
+
+    // FIXME: This function is only for testing purposes!!!
+    function set_confirmation_work_threshold(uint256 threshold) public onlyOwner {
+        confirmation_work_threshold = threshold;
+    }
+
+    function rotate_checkpoints(bytes32 new_checkpoint) public onlyOwner {
+        // drop hash from [0] position, shift rest of hashes and write new checkpoint to the [4] position
+        checkpoints[0] = checkpoints[1];
+        checkpoints[1] = checkpoints[2];
+        checkpoints[2] = checkpoints[3];
+        checkpoints[3] = checkpoints[4];
+        checkpoints[4] = new_checkpoint;
+    }
+
+    function set_tail_checkpoint(bytes32 new_checkpoint) public onlyOwner {
+        checkpoints[4] = new_checkpoint;
+    }
+
+    function get_current_checkpoints() public view returns (bytes32[] memory) {
+        return checkpoints;
+    }
+
+    function executeCommonProofProcessingLogic(SphinxProofFixture memory fixture, bool is_fork) public {
+        uint256 offset = 0;
+        uint256 i = 0;
+
+        // it reverts execution if core verification fails, so no special handling is required
+        this.verifyProof(fixture.vkey, fixture.publicValues, fixture.proof);
+
+        bytes memory confirmation_work = new bytes(32);
+        for (i = 0; i < 32; i++) {
+            confirmation_work[i] = fixture.publicValues[i];
+        }
+        offset += 32;
+
+        bytes memory first_layer_hash = new bytes(32);
+        for (i = 0; i < 32; i++) {
+            first_layer_hash[i] = fixture.publicValues[i + offset];
+        }
+        offset += 32;
+
+        bytes memory target_layer_hash = new bytes(32);
+        for (i = 0; i < 32; i++) {
+            target_layer_hash[i] = fixture.publicValues[i + offset];
+        }
+        offset += 32;
+
+        if (uint256(bytes32(confirmation_work)) < confirmation_work_threshold) {
+            revert ErrorConfirmationWorkIsSmallerThanThreshold();
+        }
+
+        if (bytes32(first_layer_hash) != checkpoints[0]) {
+            revert ErrorUnexpectedLayerHash();
+        }
```

Comment:
I don't think is is right. In a normal case it is likely that the `first_layer_hash` is actually the last index in you implemented logic. For forks, it is possible for the `first_layer_hash` to be any of the checkpoints in the current array.

To check if the array contains `first_layer_hash` you can either create a `view` function that loops through the array and look at the element (linear complexity, but we are not storing many elements) OR create a mapping `mapping(bytes32 => bool) checkpointExists` where you would toggle the boolean when different checkpoints are set/unset.

Reply:
I believe looping over the array uses less storage and might be cheaper gas since the array is so small, but either works

Reply:
I agree